pyfr/plugins/integrate.py

```diff
@@ -0,0 +1,156 @@
+        self.elminfo = []
+        # Iterate over the element types
+        for ename, eles in intg.system.ele_map.items():
+            # Locations of each solution point
+            plocupts = eles.ploc_at_np('upts')
+            
+            # Jacobians
+            rjacs = 1/eles.rcpdjac_at_np('upts')
+            
+            # Weights
+            rname = self.cfg.get('solver-elements-' + ename, 'soln-pts')
+            wts = get_quadrule(ename, rname, eles.nupts).wts
+            
+            # Save
+            self.elminfo.append((plocupts, rjacs, wts))
+        
+
+    def _init_gradients(self, intg):
+        # Determine what gradients, if any, are required
+        self._gradpnames = gradpnames = set()
+        for k, ex in self.exprs:
+            gradpnames.update(re.findall(r'\bgrad_(.+?)_[xyz]\b', ex))
+
+        # If gradients are required then form the relevant operators
+        if gradpnames:
+            self._gradop, self._rcpjact = [], []
+
+            for eles in intg.system.ele_map.values():
+                self._gradop.append(eles.basis.m4)
+
+                # Get the smats at the solution points
+                smat = eles.smat_at_np('upts').transpose(2, 0, 1, 3)
+
+                # Get |J|^-1 at the solution points
+                rcpdjac = eles.rcpdjac_at_np('upts')
+
+                # Product to give J^-T at the solution points
+                self._rcpjact.append(smat*rcpdjac)
+
+    def _eval_exprs(self, intg):
+
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        exprs = np.zeros(len(self.exprs))
+        # Iterate over each element type in the simulation
+        for i, (soln, (plocs, rjacs, wts)) in enumerate(zip(intg.soln, self.elminfo)):
+    
+            # Convert from conservative to primitive variables
+            psolns = self.elementscls.con_to_pri(soln.swapaxes(0, 1),
+                                                 self.cfg)
+
+            # Prepare the substitutions dictionary
+            subs = dict(zip(pnames, psolns))
+            subs.update(zip('xyz', plocs.swapaxes(0, 1)))
+
+            # Compute any required gradients
+            if self._gradpnames:
+                # Gradient operator and J^-T matrix
+                gradop, rcpjact = self._gradop[i], self._rcpjact[i]
+                nupts = gradop.shape[1]
+
+                for pname in self._gradpnames:
+                    psoln = subs[pname]
+
+                    # Compute the transformed gradient
+                    tgradpn = gradop @ psoln
+                    tgradpn = tgradpn.reshape(self.ndims, nupts, -1)
+
+                    # Untransform this to get the physical gradient
+                    gradpn = np.einsum('ijkl,jkl->ikl', rcpjact, tgradpn)
+                    gradpn = gradpn.reshape(self.ndims, nupts, -1)
+
+                    for dim, grad in zip('xyz', gradpn):
+                        subs['grad_{0}_{1}'.format(pname, dim)] = grad
```

Comment:
Can use new f-string syntax here.

Reply:
again taken directly from `tavg.py`, so it might be best to keep the syntax the same

Reply:
Use the cleaner syntax in new code (we're gradually switching over as code is updated/added).